pyqasm/visitor.py

```diff
@@ -759,6 +761,69 @@ def _visit_custom_gate_operation(
 
         return result
 
+    def _visit_external_gate_operation(
+        self, operation: qasm3_ast.QuantumGate, inverse: bool = False
+    ) -> list[qasm3_ast.QuantumGate]:
+        """Visit an external gate operation element.
+
+        Args:
+            operation (qasm3_ast.QuantumGate): The external gate operation to visit.
+            inverse (bool): Whether the operation is an inverse operation. Defaults to False.
+
+                            If True, the gate operation is applied in reverse order and the
+                            inverse modifier is appended to each gate call.
+                            See https://openqasm.com/language/gates.html#inverse-modifier
+                            for more clarity.
+
+        Returns:
+            list[qasm3_ast.QuantumGate]: The quantum gate that was collected.
+        """
+
+        logger.debug("Visiting custom gate operation '%s'", str(operation))
+        gate_name: str = operation.name.name
+
+        if gate_name not in self._custom_gates:
+            raise_qasm3_error(
+                f"Gate '{gate_name}' is not a custom gate and can not be marked as external.",
+                span=operation.span,
+            )
+
+        gate_definition: qasm3_ast.QuantumGateDefinition = self._custom_gates[gate_name]
```

Comment:
If a basic gate is detected, we just offload the computation to the visit basic gate method then.